packages/utils/src/dal/mikro-orm/mikro-orm-repository.ts

```diff
@@ -333,91 +345,186 @@ export function mikroOrmBaseRepositoryFactory<T extends object = object>(
+        relation.reference === ReferenceType.ONE_TO_ONE
+      ) {
+        return undefined
+      }
 
-        const existingEntity = existingEntitiesMap.get(key)
-        if (existingEntity) {
-          const updatedType = manager.assign(existingEntity, data_)
-          updatedEntities.push(updatedType)
-        } else {
-          const newEntity = manager.create<T>(entity, data_)
-          createdEntities.push(newEntity)
+      let normalizedData = []
+      if (dataForRelation.isInitialized()) {
+        normalizedData = dataForRelation.getItems()
+      } else {
+        await dataForRelation.init()
+        normalizedData = dataForRelation.getItems()
+      }
+
+      if (relation.reference === ReferenceType.MANY_TO_MANY) {
+        let upserted = []
+        if (normalizedData.length) {
+          upserted = await manager.upsertMany(normalizedData)
         }
-      })
 
-      if (createdEntities.length) {
-        manager.persist(createdEntities)
-        upsertedEntities.push(...createdEntities)
+        const currentPivotColumn = relation.inverseJoinColumns[0]
+        const parentPivotColumn = relation.joinColumns[0]
+
+        // TODO: Currently we assume that IDs are used in the pivot table, adjust this to support composite keys
+        const pivotData = upserted
+          .map((currModel) =>
+            (currModel[relation.inversedBy] as any).map((inverseModel) => {
+              return {
+                [parentPivotColumn]: inverseModel.id,
+                [currentPivotColumn]: (currModel as any).id,
+              }
+            })
+          )
+          .flat()
+
+        const qb = manager.qb(relation.pivotEntity)
+        await qb.insert(pivotData).onConflict().ignore().execute()
+
+        await manager.nativeDelete(relation.pivotEntity, {
+          [currentPivotColumn]: {
+            $nin: pivotData.map((d) => d[currentPivotColumn]),
+          },
+          [parentPivotColumn]: {
+            $in: pivotData.map((d) => d[parentPivotColumn]),
+          },
+        })
+
+        return upserted
       }
 
-      if (updatedEntities.length) {
-        manager.persist(updatedEntities)
-        upsertedEntities.push(...updatedEntities)
+      if (relation.reference === ReferenceType.ONE_TO_MANY) {
+        let upserted = []
+        if (normalizedData.length) {
+          upserted = await manager.upsertMany(normalizedData)
+        }
+
+        const joinColumns =
+          relation.targetMeta?.properties[lowerCaseFirst(entity.name)]
+            ?.joinColumns
+        const joinColumnsConstraints = {}
+        joinColumns?.forEach((joinColumn, index) => {
+          const referencedColumnName = relation.referencedColumnNames[index]
+          joinColumnsConstraints[joinColumn] = data[referencedColumnName]
+        })
+
+        // TODO: We are assuming that the table uses an ID as a primary key, adjust this to support composite keys
+        await manager.nativeDelete(relation.type, {
```

Comment:
Here we should soft delete and maybe only execute this with a explicit flag. Like having another method `upsertAndDelete`.
The reason is that the upsert (+ delete) method will have to return all the keys affected so that we can emit events + invoke the remote link about deleted entities/ids.

cc: @adrien2p

Reply:
yes we discussed that, I suggested upsertAndDelete but oli prefer `set` or something like that instead, we discussed on friday. Also, we ve talked about it with stev, and yes we will return everything impacted and how it was impacted when the feature is done. it should be farly simple because we are limiting to one level deep for relation

Reply:
yes exactly, that was my argument too, we are expecting upsert to only do create/update, so havaing upsertAndDelete would be ideal IMO